### PR TITLE
#11 Make the TVrank command-line interface more convenient

### PR DESCRIPTION
**Pull request by Olsi-B**

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -104,23 +104,11 @@ fn create_project() -> Res<ProjectDirs> {
...
-  verbose: u8,
-
-  /// Force updating internal databases.
-  #[structopt(short, long)]
-  force_update: bool,
-
-  /// Sort by year/rating/title instead of rating/year/title
-  #[structopt(short = "y", long)]
-  sort_by_year: bool,
-
```
Comment (fredmorcos) [inline src/main.rs:118]: I would prefer to keep this in the form of
```rust
#[structopt(flatten)]
opts: Opts,
```

And give precedence to options passed as subcommands over these.

File: src/main.rs [hunk 2]
```diff
@@ -129,19 +117,43 @@ enum Command {
...
+
+  /// Sort by year/rating/title instead of rating/year/title
+  #[structopt(short = "y", long)]
+  sort_by_year: bool,
+
+  /// Verbose output (can be specified multiple times)
+  #[structopt(short, long, parse(from_occurrences))]
+  verbose: u8,
+}
+
```
Comment (fredmorcos) [inline src/main.rs:156]: Just a nitpick, could you move `Opts` to be above `Opt` please?

File: src/main.rs [hunk 3]
```diff
@@ -400,10 +412,12 @@ fn run(opt: Opt) -> Res<()> {
...
-  match opt.command {
-    Command::Title { exact, title } => imdb_single_title(&title, &imdb, &imdb_url, opt.sort_by_year, exact)?,
-    Command::MoviesDir { dir } => imdb_movies_dir(&dir, &imdb, &imdb_url, opt.sort_by_year)?,
-    Command::SeriesDir { dir } => imdb_series_dir(&dir, &imdb, &imdb_url, opt.sort_by_year)?,
+  match cmd {
+    Command::Title { exact, title, opts } => {
+      imdb_single_title(&title, &imdb, &imdb_url, opts.sort_by_year, exact)?
+    }
+    Command::MoviesDir { dir, opts } => imdb_movies_dir(&dir, &imdb, &imdb_url, opts.sort_by_year)?,
+    Command::SeriesDir { dir, opts } => imdb_series_dir(&dir, &imdb, &imdb_url, opts.sort_by_year)?,
```
Comment (fredmorcos) [inline src/main.rs:420]: I guess the precedence would be implemented here, something like:

```rust
let sort_by_year = opts.sort_by_year || opt.sort_by_year;
...
```

File: src/main.rs [hunk 4]
```diff
@@ -416,8 +430,13 @@ fn run(opt: Opt) -> Res<()> {
 fn main() {
   let start_time = Instant::now();
   let opt = Opt::from_args();
+  let opts = match opt.command.clone() {
```
Comment (fredmorcos) [inline src/main.rs:433]: I feel like it would be possible to avoid the `clone()` call here and remove the `Clone` derives on `Opts` and `Command`.

**Review by fredmorcos — COMMENTED** [review]
Thank you so much for this contribution, it's awesome! 😃 

A few points, apart from the code review comments:
- [ ] You can now rebase on fredmorcos/tvrank/main to fix the CI lint issue.
- [ ] After the proposed changes, you can update the README by mentioning that some options can come before or after a command and that the latter gets precedence.

**Comment by Olsi-B** [comment]
PR updated. Options precedence is now in `main` function resolved.

File: src/main.rs [hunk 5]
```diff
@@ -416,8 +433,23 @@ fn run(opt: Opt) -> Res<()> {
...
+  };
+
+  let merge_opts = Opts {
+    force_update: opts.force_update || opt.opts.force_update,
+    sort_by_year: opts.sort_by_year || opt.opts.sort_by_year,
+    verbose: if opts.verbose > 0 {
+      opts.verbose
+    } else {
+      opt.opts.verbose
+    },
```
Comment (fredmorcos) [inline src/main.rs:449]: I think this should perhaps be `verbose: opts.verbose + opt.opts.verbose`? What do you think?

File: src/main.rs [hunk 6]
```diff
@@ -416,8 +433,23 @@ fn run(opt: Opt) -> Res<()> {
 fn main() {
   let start_time = Instant::now();
   let opt = Opt::from_args();
+  let opts = match &opt.command {
+    Command::Title { exact: _, title: _, opts } => opts,
+    Command::MoviesDir { dir: _, opts } => opts,
+    Command::SeriesDir { dir: _, opts } => opts,
```
Comment (fredmorcos) [inline src/main.rs:439]: Any reason we need the `: _`s here?

File: src/main.rs [hunk 7]
```diff
@@ -400,10 +415,12 @@ fn run(opt: Opt) -> Res<()> {
...
-  match opt.command {
-    Command::Title { exact, title } => imdb_single_title(&title, &imdb, &imdb_url, opt.sort_by_year, exact)?,
-    Command::MoviesDir { dir } => imdb_movies_dir(&dir, &imdb, &imdb_url, opt.sort_by_year)?,
-    Command::SeriesDir { dir } => imdb_series_dir(&dir, &imdb, &imdb_url, opt.sort_by_year)?,
+  match cmd {
+    Command::Title { exact, title, opts: _ } => {
+      imdb_single_title(&title, &imdb, &imdb_url, opt.sort_by_year, exact)?
+    }
+    Command::MoviesDir { dir, opts: _ } => imdb_movies_dir(&dir, &imdb, &imdb_url, opt.sort_by_year)?,
+    Command::SeriesDir { dir, opts: _ } => imdb_series_dir(&dir, &imdb, &imdb_url, opt.sort_by_year)?,
```
Comment (fredmorcos) [inline src/main.rs:423]: Also here, any reason we need the `: _`s?

File: README.md [hunk 8]
```diff
@@ -92,6 +92,14 @@ level:
 $ tvrank -vvv --sort-by-year title "city of god"
 ```
 
+The following options can come before or after the sub-command. The latter have precedence over the former.
+
+```sh
+-v
```
Comment (fredmorcos) [inline README.md:98]: Perhaps change this one to `--verbose`?

File: src/main.rs (hunk 5 above)
Comment (Olsi-B) [reply src/main.rs:449]: `verbose` is a number, not a boolean. If we do this, then  an example with `-vvv` as first option and `-v` as second would give `-vvvv`, which is `Debug`(level 4). We want to overwrite the first(level 3) with the second(level 1) instead.

Comment (fredmorcos) [reply src/main.rs:449]: Alright!

**Comment by fredmorcos** [comment]
This is really awesome! Thank you again 👍🏽
It should get merged once the CI is done

**Comment by fredmorcos** [comment]
@Olsi-B Can you just leave a comment on issue #11 so that I set you as the person assigned to that issue?

**Pull request merged**